docs/design/operator.md

```diff
@@ -0,0 +1,200 @@
+configuration parameters. This configuration will allow for extensive
+customization if desired but will assume sane defaults for as many parameters as
+possible to allow for as minimal a configuration as possible.
+
+**Target Configurations:** Hosted, Managed
+
+### Import
+
+The Gluster operator will be able to import an existing Gluster cluster at any
+stage of initialization. The operator's desired configuration for the cluster
+will need to be compatible with the state of the cluster at the time of import.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Cluster Upgrade
+
+The Gluster operator will be able to execute a rolling upgrade of any hosted or
+managed Gluster cluster. It will trigger an upgrade one node at a time, waiting
+for the node to come up and for the cluster to report no pending heals to any
+files before triggering a subsequent node.
+
+**Target Configurations:** Hosted, Managed
+
+### Monitoring
+
+To provide event-driven features, the Gluster operator will monitor various
+aspects of the clusters it will automate.
+
+#### Storage Capacity
+
+The Gluster operator will monitor the storage capacity of the overall cluster.
+Storage capacity is defined here as the amount of space that can be committed to
+new Gluster volumes and not how much free space is available. It will trigger
+events when storage capacity reaches "Warning" levels and "Critical" levels.
+These levels will be configurable, and should be set to a value that will allow
+for corrective action to take place before the storage reaches maximum capacity.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Volume Density
+
+The Gluster operator will monitor the total volume density of any cluster it is
+automating. It will trigger events when the number of volumes in the cluster
+reaches a pre-defined level that should prompt corrective action. This value
+will be configurable, and should be set to a value that will allow for
+corrective action to take place before cluster nodes lag or crash.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Device Health
+
+The Gluster operator will monitor the health of all devices in an automated
+Gluster cluster. It will report a "Warning" when a device reports failing health
+and will report a "Critical" problem when a device becomes inaccessible. Support
+for this feature may vary form one device type to another.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Volume Health
+
+The Gluster operator will monitor the health of all volumes in an automated
+Gluster cluster. It will report a "Warning" when a volume reports failing bricks
+and will report a "Critical" problem when a volume becomes inaccessible.
+
+**Target Configurations:** External, Hosted, Managed
+
+#### Node Health
+
+The Gluster operator will monitor the health of all nodes in an automated
+Gluster cluster. It will report a "Critical" problem when a node becomes
+inaccessible.
```

Comment:
I'll second @vbellur request to monitor additional services. Currently the readiness probe only checks `glusterd`, while `tcmu-runner` `gluster-block-target` and `gluster-blockd` are all critical to a healthy gluster pod. Perhaps that just needs to be pushed into the readiness probe though?

Reply:
Would this also include the other pods like `glusterblock-provisioner` and `heketi` ?

Reply:
> Perhaps that just needs to be pushed into the readiness probe though?

Yes. :)

> Would this also include the other pods like `glusterblock-provisioner` and `heketi` ?

That's implementation-specific. Neither of those two are guaranteed to always be around.

Reply:
> >  Would this also include the other pods like glusterblock-provisioner and heketi ?
> 
> That's implementation-specific. Neither of those two are guaranteed to always be around.

But I think in the current situation, they would certainly belong to the scope of the operator!

Reply:
I think this point is valid though. The document reads as if the operator only cares about gluster nodes. But maybe we need to define what are the "nodes in an automated Gluster cluster". Those will be gluster nodes, but possibly other pods in kubernetes. At least currently it will include heketi and (optionally) a gluster-block-provisioner pod. Etc.